File: sdk/python/bunqueue/queue_query.py
"""Query operations mixin for Queue (jobs, states, counts, logs, children)."""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Sequence, Union

from .connection import _compact
from .errors import CommandError, CommandTimeoutError
from .job import Job


def _values_payload(response: Dict[str, Any]) -> Dict[str, Any]:
    """Unwrap {data: {values: {...}}} (server) or top-level values."""
    data = response.get("data")
    if isinstance(data, dict) and "values" in data:
        return dict(data.get("values") or {})
    return dict(response.get("values") or {})


class QueueQueryOps:
    """Mixin: requires ``self.name`` and ``self.connection``."""

    # ------------------------------------------------------------------ jobs

    def get_job(self, job_id: str) -> Optional[Job]:
        try:
            response = self.connection.call({"cmd": "GetJob", "id": job_id})
        except CommandError as exc:
            if "not found" in str(exc).lower():
                return None
            raise
        raw = response.get("job")
        return Job(raw, self.connection) if raw else None

    def get_job_by_custom_id(self, custom_id: str) -> Optional[Job]:
        try:
            response = self.connection.call({"cmd": "GetJobByCustomId", "customId": custom_id})
        except CommandError as exc:
            if "not found" in str(exc).lower():
                return None
            raise
        raw = response.get("job")
        return Job(raw, self.connection) if raw else None

    def get_jobs(
        self,
        state: Optional[Union[str, Sequence[str]]] = None,
        start: int = 0,
        end: int = 1000,
    ) -> List[Job]:
        """Mirror TS getJobsAsync: offset=start, limit=end-start."""
        command = _compact(
            {
                "cmd": "GetJobs",
                "queue": self.name,
                "state": list(state) if isinstance(state, (list, tuple)) else state,
                "offset": start,
                "limit": max(end - start, 0),
            }
        )
        response = self.connection.call(command)
        return [Job(raw, self.connection) for raw in response.get("jobs", [])]

    def get_waiting(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("waiting", start, end)

    def get_delayed(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("delayed", start, end)

    def get_active(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("active", start, end)

    def get_completed(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("completed", start, end)

    def get_failed(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("failed", start, end)

    def get_prioritized(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("prioritized", start, end)

    def get_waiting_children(self, start: int = 0, end: int = 1000) -> List[Job]:
        return self.get_jobs("waiting-children", start, end)

    # ----------------------------------------------------------------- state

    def get_state(self, job_id: str) -> str:
        response = self.connection.call({"cmd": "GetState", "id": job_id})
        return str(response.get("state"))

    def get_result(self, job_id: str) -> Any:
        response = self.connection.call({"cmd": "GetResult", "id": job_id})
        return response.get("result")

    def get_progress(self, job_id: str) -> Dict[str, Any]:
        response = self.connection.call({"cmd": "GetProgress", "id": job_id})
        return {"progress": response.get("progress"), "message": response.get("message")}

    def wait_for_job(self, job_id: str, timeout_ms: int = 30000) -> Any:
        """Block until the job completes; returns its result.

        The server's ``WaitJob`` waiter resolves only on completion, answering
        ``{ok:true, completed:false}`` (no result) otherwise — so returning
        ``None`` would be indistinguishable from a genuine ``None`` result. On
        non-completion we probe the job state: a ``failed`` job raises
        :class:`CommandError` (it will not complete), everything else raises
        :class:`CommandTimeoutError`."""
        response = self.connection.call(
            {"cmd": "WaitJob", "id": job_id, "timeout": timeout_ms},
            timeout=timeout_ms / 1000 + 5,
        )
        if not response.get("completed"):
            try:
                state = self.get_state(job_id)
            except CommandError:
                state = None
            if state == "failed":
                raise CommandError(f"job {job_id} failed before completion")
            raise CommandTimeoutError(f"waitUntilFinished timed out after {timeout_ms}ms")
        return response.get("result")

    def wait_job_until_finished(
        self, job_id: str, queue_events: Any = None, ttl_ms: Optional[int] = None
    ) -> Any:
        """BullMQ-parity alias for :meth:`wait_for_job` (`queue_events` unused over TCP)."""
        del queue_events
        return self.wait_for_job(job_id, ttl_ms if ttl_ms is not None else 30000)

    # -------------------------------------------------------------- children

    def get_children_values(self, job_id: str) -> Dict[str, Any]:
        response = self.connection.call({"cmd": "GetChildrenValues", "id": job_id})
        return _values_payload(response)

    def get_failed_children_values(self, job_id: str) -> Dict[str, Any]:
        response = self.connection.call({"cmd": "GetFailedChildrenValues", "id": job_id})
        return _values_payload(response)

    def get_ignored_children_failures(self, job_id: str) -> Dict[str, Any]:
        response = self.connection.call({"cmd": "GetIgnoredChildrenFailures", "id": job_id})
        return _values_payload(response)

    def remove_child_dependency(self, job_id: str) -> None:
        self.connection.call({"cmd": "RemoveChildDependency", "id": job_id})

    def remove_unprocessed_children(self, job_id: str) -> None:
        self.connection.call({"cmd": "RemoveUnprocessedChildren", "id": job_id})

    def get_deduplication_job_id(self, deduplication_id: str) -> Optional[str]:
        """Id of the job currently holding this deduplication key (customId)."""
        job = self.get_job_by_custom_id(deduplication_id)
        return job.id if job else None

    # ---------------------------------------------------------------- counts

    def get_job_counts(self) -> Dict[str, int]:
        response = self.connection.call({"cmd": "GetJobCounts", "queue": self.name})
        return dict(response.get("counts", {}))

    def get_waiting_count(self) -> int:
        return self.get_job_counts().get("waiting", 0)

    def get_active_count(self) -> int:
        return self.get_job_counts().get("active", 0)

    def get_completed_count(self) -> int:
        return self.get_job_counts().get("completed", 0)

    def get_failed_count(self) -> int:
        return self.get_job_counts().get("failed", 0)

    def get_delayed_count(self) -> int:
        return self.get_job_counts().get("delayed", 0)

    def get_prioritized_count(self) -> int:
        return self.get_job_counts().get("prioritized", 0)

    def get_waiting_children_count(self) -> int:
        return self.get_job_counts().get("waiting-children", 0)

    def count(self) -> int:
        response = self.connection.call({"cmd": "Count", "queue": self.name})
        return int(response.get("count", 0))

    def get_counts_per_priority(self) -> Dict[str, int]:
        response = self.connection.call({"cmd": "GetCountsPerPriority", "queue": self.name})
        return dict(response.get("counts") or response.get("data") or {})

    # ------------------------------------------------------------------ logs

    def add_job_log(self, job_id: str, message: str, level: Optional[str] = None) -> None:
        self.connection.call(_compact({"cmd": "AddLog", "id": job_id, "message": message, "level": level}))

    def get_job_logs(self, job_id: str, start: Optional[int] = None, end: Optional[int] = None) -> List[Any]:
        response = self.connection.call(_compact({"cmd": "GetLogs", "id": job_id, "start": start, "end": end}))
        data = response.get("data")
        if isinstance(data, dict):
            return list(data.get("logs") or [])
        return list(response.get("logs") or [])

    def clear_job_logs(self, job_id: str, keep_logs: Optional[int] = None) -> None:
        self.connection.call(_compact({"cmd": "ClearLogs", "id": job_id, "keepLogs": keep_logs}))

File: sdk/elixir/lib/bunqueue/queue_query.ex
defmodule Bunqueue.QueueQuery do
  @moduledoc false

  alias Bunqueue.{CommandError, Job, Queue, TimeoutError}

  def get_job(queue, id) do
    not_found_nil(fn ->
      with {:ok, response} <- Queue.call(queue, %{"cmd" => "GetJob", "id" => id}) do
        {:ok, from_response(response["job"], queue)}
      end
    end)
  end

  def get_job_by_custom_id(queue, custom_id) do
    not_found_nil(fn ->
      command = %{"cmd" => "GetJobByCustomId", "queue" => queue.name, "customId" => custom_id}

      with {:ok, response} <- Queue.call(queue, command) do
        {:ok, from_response(response["job"], queue)}
      end
    end)
  end

  def get_jobs(queue, state \\ "waiting", offset \\ 0, limit \\ 100) do
    command = %{
      "cmd" => "GetJobs",
      "queue" => queue.name,
      "state" => state,
      "offset" => max(offset, 0),
      "limit" => max(limit, 0)
    }

    with {:ok, response} <- Queue.call(queue, command) do
      jobs = Enum.map(response["jobs"] || [], &Job.from_wire(&1, queue.connection))
      {:ok, jobs}
    end
  end

  def get_state(queue, id), do: field(queue, %{"cmd" => "GetState", "id" => id}, "state")
  def get_result(queue, id), do: field(queue, %{"cmd" => "GetResult", "id" => id}, "result")

  def get_progress(queue, id) do
    with {:ok, response} <- Queue.call(queue, %{"cmd" => "GetProgress", "id" => id}) do
      {:ok, %{progress: response["progress"], message: response["message"]}}
    end
  end

  def get_job_counts(queue) do
    field(queue, %{"cmd" => "GetJobCounts", "queue" => queue.name}, "counts", %{})
  end

  def count(queue), do: field(queue, %{"cmd" => "Count", "queue" => queue.name}, "count", 0)

  def wait_for_job(queue, id, timeout_ms) do
    timeout = timeout_ms |> normalize_timeout() |> max(0) |> min(600_000)
    command = %{"cmd" => "WaitJob", "id" => id, "timeout" => timeout}

    with {:ok, response} <- Queue.call(queue, command, timeout + 5_000) do
      if response["completed"] do
        {:ok, response["result"]}
      else
        wait_timeout(queue, id, timeout)
      end
    end
  end

  def get_logs(queue, id, start \\ nil, finish \\ nil) do
    command = compact(%{"cmd" => "GetLogs", "id" => id, "start" => start, "end" => finish})

    with {:ok, response} <- Queue.call(queue, command) do
      {:ok, get_in(response, ["data", "logs"]) || []}
    end
  end

  defp wait_timeout(queue, id, timeout) do
    case get_state(queue, id) do
      {:ok, "failed"} -> {:error, %CommandError{message: "job #{id} failed", command: "WaitJob"}}
      _ -> {:error, %TimeoutError{message: "WaitJob timed out", timeout: timeout}}
    end
  end

  defp field(queue, command, key, default \\ nil) do
    with {:ok, response} <- Queue.call(queue, command), do: {:ok, Map.get(response, key, default)}
  end

  defp from_response(nil, _queue), do: nil
  defp from_response(raw, queue), do: Job.from_wire(raw, queue.connection)

  defp not_found_nil(function) do
    case function.() do
      {:error, %CommandError{} = error} ->
        if CommandError.not_found?(error), do: {:ok, nil}, else: {:error, error}

      result ->
        result
    end
  end

  defp compact(map), do: Map.reject(map, fn {_key, value} -> is_nil(value) end)
  defp normalize_timeout(value) when is_number(value), do: trunc(value)
  defp normalize_timeout(_value), do: 0
end
